app/liquid_tags/user_subscription_tag.rb

```diff
@@ -1,9 +1,242 @@
+        'X-CSRF-Token': window.csrfToken,
+        'Content-Type': 'application/json',
+      }
+
+      const articleId = document.getElementById('article-body').dataset.articleId;
+      const subscriber = userData();
+      const body = JSON.stringify(
+          {
+            user_subscription: {
+              source_type: 'Article',
+              source_id: articleId,
+              subscriber_email: subscriber.email
+            }
+          }
+        )
+
+      return fetch('/user_subscriptions', {
+        method: 'POST',
+        headers: headers,
+        credentials: 'same-origin',
+        body: body,
+      }).then(function(response) {
+        return response.json();
+      });
+    }
+
+    function fetchIsSubscribed() {
+      const articleId = document.getElementById('article-body').dataset.articleId;
+
+      const params = new URLSearchParams({
+        source_type: 'Article',
+        source_id: articleId
+      }).toString();
+
+      const headers = {
+        Accept: 'application/json',
+        'X-CSRF-Token': window.csrfToken,
+        'Content-Type': 'application/json',
+      }
+
+      return fetch(`/user_subscriptions/subscribed?${params}`, {
+        method: 'GET',
+        headers: headers,
+        credentials: 'same-origin',
+      }).then(function(response) {
+        if (response.ok) {
+          return response.json();
+        } else {
+          console.error(`Base data error: ${response.status} - ${response.statusText}`);
+        }
+      });
+    }
+
+    // Handle API responses
+    // ***************************************
+    function handleSubscription() {
+      submitSubscription().then(function(response) {
+        if (response.success) {
+          const authorUsername = document.getElementById('user-subscription-tag').dataset.authorUsername;
+          const successMsg = `You are now subscribed and may receive emails from ${authorUsername}`;
+          showResponseMessage('success', successMsg);
+        } else {
+          showResponseMessage('danger', response.error);
+        }
+      });
+    }
+
+    function checkIfSubscribed() {
+      fetchIsSubscribed().then(function(response) {
+        const subscriber = userData();
+        const isSubscriberAuthedWithApple = subscriber.email.endsWith('@privaterelay.appleid.com');
+
+        if (response.is_subscribed) {
+          showSubscribed();
+        } else if (isSubscriberAuthedWithApple) {
+          showAppleAuthMessage();
+        } else {
+          updateSubscriberData();
+        }
+      });
+    }
+
+    // The markup defaults to signed out UX
+    if (isUserSignedIn()) {
+      showSignedIn();
+      addConfirmationModalClickHandlers();
+
+      // We need access to some DOM elements (i.e. csrf token, article id, userData, etc.)
+      document.addEventListener('DOMContentLoaded', function() {
+        checkIfSubscribed();
+      });
+    } else {
+      addSignInClickHandler();
+    }
+  JAVASCRIPT
```

Comment:
feel free to push back on this, but i wonder if it would be worth it to extract this into a separate file and the pull it in and assign it to `SCRIPT`. not majorly important, but because there is _so much_ JS for this tag, i was thinking that it might help with the readability of this model.

Reply:
I considered moving it into its own file, but I decided against it. The JS here is treated as a raw string. I considered moving the code into a file and do something like:
```ruby
SCRIPT = File.read('app/javascripts/liquid_tags/user_subscription_tag.js').freeze

def self.script
  SCRIPT
end
```
This introduces a couple more items for discussion. It's been a long outstanding TODO for us to reorganize the JS. Also, while I would obviously add a comment to the top of this file (if this approach were taken), it wouldn't be clear that this file would sort of fall outside of the usual asset pipeline. You can't include anything in this file, for example. I felt that got a little confusing. Instead, I chose to stick with the "current setup" of how we're handling liquid tags and re-visit the entire architecture of liquid tags in the future 🙃 .

Reply:
@atsmith813 That's fair; fwiw, when I left this comment I was thinking of a `File.read` type of thing as well. I think re-visiting it potentially later is fine, I think it's 👍 to stick with convention for now.